### PR TITLE
[bugfix] Don't check for autosync on manual triggered sync (#3026)

### LINKED ISSUE #3026: autosync = false breaks manual sync
Setting `core.autosync = false` breaks manual syncing with `gopass sync` command. It then just does nothing (it seems):

    19:44:07 mcn@x2b ~ grep sync .config/gopass/config 
    autosync=false
    
    19:44:54 mcn@x2b ~ time gopass sync
    🚥 Syncing with all remotes ...
    ✅ All done
    
    real	0m0,437s
    user	0m0,270s
    sys	0m0,179s
    
In pretty much no time it "syncs" and states all being done, while actually not at all syncing. When setting `core.autosync = true` it then does it also manually:

    19:44:57 mcn@x2b ~ grep sync .config/gopass/config 
    autosync=true
    
    19:45:39 mcn@x2b ~ time gopass sync
    🚥 Syncing with all remotes ...
    [<root>] 
       gitfs pull and push ... OK (no changes)
       done
       …
    
I prefer it not autosyncing, because I find it annoying to wait for the syncing of >20 stores when I only want to use the command line tool. So I have set `autosync.interval = 9999` to work around this issue. Is there another way to do that?

gopass 1.15.15 go1.23.4 linux amd64 on NixOS unstable with `home.packages = with pkgs; [ gopass gopass-jsonapi ];`

### PR DESCRIPTION
**Pull request by TM2500**
* [bugfix] Don't check for autosync on manual triggered sync

Fixes #3026

### CODE REVIEW

**Comment by TM2500** [comment]
Additional Tests are needed to verify this bugfix.
As soon as they are created, the draft state is removed from the PR.

File: internal/action/sync.go [hunk 1]
```diff
@@ -85,6 +85,7 @@ func (s *Action) autoSync(ctx context.Context) error {
 	debug.Log("autosync - interval: %s", syncInterval)
 
 	if time.Since(ls) > syncInterval {
+		ctx = ctxutil.WithInteractive(ctx, true)
```
Comment (AnomalRoil) [inline internal/action/sync.go:88]: It looks like you're relying on the interactive context key in the end, instead of the newly added autosync one, no?

File: pkg/ctxutil/ctxutil.go [hunk 2]
```diff
@@ -489,3 +490,18 @@ func IsHidden(ctx context.Context) bool {
...
+
+// HasAutoSync returns true if syncs are triggered by autosync.
+func HasAutoSync(ctx context.Context) bool {
+	bv, ok := ctx.Value(ctxKeyAutoSync).(bool)
+	if !ok {
+		return false
+	}
+
+	return bv
+}
```
Comment (AnomalRoil) [inline pkg/ctxutil/ctxutil.go:507]: Do you really need these?
It looks like you're relying on the interactive context key above, instead of that newly added autosync one, no?

File: internal/action/sync.go [hunk 3]
```diff
@@ -161,10 +162,12 @@ func (s *Action) sync(ctx context.Context, store string) error {
 func (s *Action) syncMount(ctx context.Context, mp string) error {
 	// using GetM here to get the value for this mount, it might be different
 	// than the global value
-	if as := s.cfg.GetM(mp, "core.autosync"); as == "false" {
-		debug.Log("not syncing %s, autosync is disabled for this mount", mp)
+	if ctxutil.HasAutoSync(ctx) {
```
Comment (AnomalRoil) [inline internal/action/sync.go:165]: Wait, you're setting the interactive key above, not the autosync one, no?

File: internal/action/sync.go (hunk 1 above)
Comment (AnomalRoil) [reply internal/action/sync.go:88]: You might mean to use 
```suggestion
		ctx = ctxutil.WithAutoSync(ctx, true)
```
there, no?

File: internal/action/sync.go (hunk 3 above)
Comment (TM2500) [reply internal/action/sync.go:165]: I propably mixed something up. I need to take a look at it again.
I hope to include meaningful tests as well, as there are none for sync.

File: internal/action/sync.go (hunk 1 above)
Comment (TM2500) [reply internal/action/sync.go]: Yes, I intended to use the new context.

**Comment by TM2500** [comment]
For tests I would have to write new tests utils, that support fake-remote git repos as stores. This may take me a while to figure out how to include it in the available tooling.
The bug is fixed with the last two commits.

**Comment by dominikschulz** [comment]
@TM2500 Thanks for the PR. I won't press for tests if they don't provide more value than maintanence overhead. Feel free to mark this ready for review as is or whenever you think it's ready. Then I'll give it a good look and let you know if I think anything is amiss.

**Comment by TM2500** [comment]
> @TM2500 Thanks for the PR. I won't press for tests if they don't provide more value than maintanence overhead. Feel free to mark this ready for review as is or whenever you think it's ready. Then I'll give it a good look and let you know if I think anything is amiss.

Yes, I think tests are an order of magnitude more work, as we would have to find a way to generate dummy-remote stores in our test suite.

I will mark this for review and the PR transitions from draft to a regular PR.

Thank you.

File: internal/action/sync.go [hunk 4]
```diff
@@ -161,10 +162,12 @@ func (s *Action) sync(ctx context.Context, store string) error {
 func (s *Action) syncMount(ctx context.Context, mp string) error {
 	// using GetM here to get the value for this mount, it might be different
 	// than the global value
```
Comment (AnomalRoil) [inline internal/action/sync.go:164]: you could have moved the comment inside of the first `if` I guess

**Review by AnomalRoil — DISMISSED** [review]
LGTM

**Review by dominikschulz — COMMENTED** [review]
You are adding a bunch of code just to pass a bool in to the sync method.
Wouldn't it easier to just pass an additional bool to s.sync to signal that it's being invoked from autosync? Or is there something I'm missing?

**Comment by TM2500** [comment]
> You are adding a bunch of code just to pass a bool in to the sync method. Wouldn't it easier to just pass an additional bool to s.sync to signal that it's being invoked from autosync? Or is there something I'm missing?

You are right, as long as we only depend on the autosync-context only within the sync package, we could just use a function-parameter.
Since there is currently no need for it in other packages, I will rewrite the Patch.

**Comment by TM2500** [comment]
I refactored to function arguments.
This is ready to merge.

**Review by dominikschulz — APPROVED** [review]
Much simpler. Thanks a lot!
